src/services/keystore.ts

```diff
@@ -68,71 +79,77 @@ export class TrustedKeyStoreService {
     }
 
     /**
-     * Convert a string to an Uint8Array.
+     * DEPRECATED! Only exists in order to migrate previous storages.
      *
-     * This function is quite primitive, Unicode is not supported.
-     */
-    private stringToBytes(str: string): Uint8Array {
-        const arr = [];
-        for (let i = 0; i < str.length; i++) {
-            arr.push(str.charCodeAt(i));
-        }
-        return new Uint8Array(arr);
-    }
-
-    /**
      * Convert a password string to a NaCl key. This is done by getting a
      * SHA512 hash and returning the first 32 bytes.
      */
-    private pwToKey(password: string): Uint8Array {
-        const bytes = this.stringToBytes(password);
-        const hash = nacl.hash(bytes);
+    private deprecatedPwToKey(password: Uint8Array): Uint8Array {
+        const hash = nacl.hash(password);
         return hash.slice(0, nacl.secretbox.keyLength);
     }
 
     /**
      * Store the trusted key (and optionally the push token) in local browser
      * storage. Encrypt it using NaCl with the provided password.
      */
-    public storeTrustedKey(
-        ownPublicKey: Uint8Array,
+    public async storeTrustedKey(
         ownSecretKey: Uint8Array,
         peerPublicKey: Uint8Array,
         pushToken: string | null,
         pushTokenType: threema.PushTokenType | null,
         password: string,
         isAutoSession: boolean,
-    ): void {
+    ): Promise<void> {
+        // Generate a random salt of 32 bytes
+        const salt = nacl.randomBytes(32);
+
+        // Determine work factor (N) to take ~500ms but use at least a work factor of 2^16.
+        // Round the work factor to the nearest power of two.
```

Comment:
```suggestion
        // Round the work factor to the nearest power of two.
```

It's not the nearest and behaves more like floor.